pymeasure/instruments/channel.py

```diff
@@ -0,0 +1,103 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2022 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+
+from .common_base import CommonBase
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class Channel(CommonBase):
+    """The base class for channel definitions.
+
+    This class supports dynamic properties like :class:`Instrument`,
+    but requires an :class:`Instrument` instance as a parent for communication.
+    The default implementation of :meth:`write` uses :code:`str.format` to
+    exchange :code:`'{ch}'` for the channel name.
+
+    :param parent: The instrument (an instance of :class:`~pymeasure.instruments.Instrument`)
+        to which the channel belongs.
+    :param id: Identifier of the channel, as it is used for the communication.
+    """
+
+    def __init__(self, parent, id):
+        super().__init__()
+        self.parent = parent
+        self.id = id
+
+    # Calls to the instrument
+    def write(self, command, **kwargs):
+        """Write a string command to the instrument.
+
+        The channel id is inserted into the command and `write_termination` appended.
+
+        Subclass this method if you want to do something else,
+        like always prepending the channel id.
+
+        :param command: command string to be sent to the instrument.
+            '{ch}' is replaced by the channel id.
+        :param kwargs: Keyword arguments for the adapter.
+        """
+        self.parent.write(command.format(ch=self.id), **kwargs)
```

Comment:
Is `:CHAN{ch}` prepended to all commands? Above (before I read this) I proposed to prefix the channel-specific stuff at the channel layer, which might be a simpler solution.

I have to say, that partial string interpolation technique is neat! :heart:

Reply:
I'm thinking if/how we could achieve allowing for this advanced technique, while not making the simple cases more complicated, by hiding the complexity internally.

Reply:
What I did in the AWG 401, which has a Sequencer containing channels in turn:
The channel knows its own ID and that of its parent sequencer.
Finally the channel's `write` injects both IDs at the same time.

This keeps the core simple, as other instruments rarely need it that complicated. 

Thanks for showing this partial format technique.

Reply:
> Is `:CHAN{ch}` prepended to all commands? 

No unfortunately not, the CHAN was just for clarification, but never actually used like thisnwith this VNA I'm working on. Some examples are:
`:SOUR{ch}:POW:PORT{pt}`
`:CALC{ch}:PAR{tr}:SEL`
`:SENS{ch}:SWE:CW`

> Finally the channel's `write` injects both IDs at the same time.
 
Huh, that's also a very interesting approach, didn't think of that yet.

Reply:
> This keeps the core simple, as other instruments rarely need it that complicated. 

I agree by the way that we should consider if such complication is worth it; I do not have a very good view on him many systems with channels will have nested channels. Right now the system you have implemented (AWG401) and the one I'm working on do have nested channels, but I can imagine that this is not very common.

Reply:
Keeping the core simple would be attractive! We can at least document the approach of the AWG. 

Wild idea (the evening is late and I might as well be wrong), but if a user that needs to do the "advanced mode" (say with three layers, where it slowly gets complicated to have one place know all necessary tokens), is it enough to override the `format_command` method in those layers to implement the `format_map` approach? If yes, we could just document that in an "advanced" section, and the _core still stays simple_!
Casper, could you assess/prototype the feasibility of that approach with your VNA?

By the way, the most complicated instrument I remember recently is the [B1500](https://pymeasure.readthedocs.io/en/latest/api/instruments/agilent/agilentB1500.html)

Reply:
> Casper, could you assess/prototype the feasibility of that approach with your VNA?

I tried and it seems to work. 

Additionally, I just found something interesting to using `format_map` rather than the normal `format` in general: apparently its a bit faster (see [this](https://stackoverflow.com/a/36780792) explanation), and you don't per se need the `PartialFormatDict`: if you format the command in your nested channel like this `"SOUR{{ch}}:POW:PORT{pt}?"` (note the double `{{` and `}}` around ch), format_map(pt=1) will return `"SOUR{ch}:POW:PORT1?"` (note the single `{` and `}` around ch). We could use this to a) keep the core simple for simple devices (because I don't see a reason not to use `format_map` rather than `format`) and b) not have to do anything special (except document this approach) for more complex, nested channels. Only thing we need then is the following `replace_placeholder` method and a `placeholder` attribute:

``` python
    def replace_placeholder(self, command):
        return command.format_map({self.placeholder: self.id})
```

Reply:
Oh, apparently `"SOUR{{ch}}:POW:PORT{pt}?".format(pt=1)` also give `"SOUR{ch}:POW:PORT1?"`.

Reply:
> edit 2: It does seem that for deeper nested channels the amount of brackets increases fast (it halves the amount of nested brackets everytime format or format_map is called): 1 bracket for level 1, 2 for level 2, 4 for level 3, 8 for level 4, 16 for level 5; I'm not sure if this is an issue?

I don't think, that it is an issue. Having that many levels is an edge case of an instrument (and cumbersome to begin with). For three levels it's still ok and maybe one of the levels has a fixed prefix or something else, such that it reduces the effective level to 2.

In the explanation you linked, is another sweet usage of `format_map`, where you pass a class, which asks the user for input for every key. Thank you very much for showing `format_map`.

Reply:
I'm not happy with `replace_placeholder`, as you might prepend always the id (such that you don't have any placeholder) in some instruments. The following function (in a subclass) does not seem right to me (as there is no placeholder anymore):
```
replace_placeholder(self, command):
    return self.id + ":" + command
```
What do you think about: `insert_id` (as you insert the id into the command, either via format_map or string addition...) or `format_command` (which was already rejected, but fits better both options)?

Reply:
Your right `replace_placeholder` doesn't really work for the prefix case. Either option is fine for me; `insert_id` seems most clear as to what it does, although one could argue that prepending a prefix is more than 'just' inserting an id.